File: backend/components/metiers/threed/tests_centralises/utils/helpers/fallback/test_fallback_helper_advanced.py
# test_fallback_helper_advanced.py – Tests ultra avancés pour fallback.py (helpers/fallback)
import pytest
from backend.components.metiers.threed.utils.logger.fallback.fallback import logger_fallback
from backend.components.metiers.threed.utils.helpers.fallback.fallback import fallback_helper

def test_logger_fallback_auditabilite():
    log = logger_fallback('Fallback test')
    assert 'timestamp' in log
    assert log['message'] == 'Fallback test'

def test_logger_fallback_vide():
    log = logger_fallback('')
    assert 'timestamp' in log
    assert log['message'] == ''

def test_fallback_helper_success():
    assert fallback_helper(lambda: 42, 0) == 42

def test_fallback_helper_default():
    assert fallback_helper(lambda: (_ for _ in ()).throw(Exception('fail')), 'defaut') == 'defaut'

def test_fallback_helper_null():
    assert fallback_helper(lambda: None, 'x') is None

def test_fallback_helper_type():
    assert isinstance(fallback_helper(lambda: 'ok', 'ko'), str)

File: backend/components/metiers/threed/tests_centralises/utils/helpers/fallback/test_fallback_helper_advanced.js
// test_fallback_helper_advanced.js – Tests ultra avancés pour fallback.js (helpers/fallback)
const { loggerFallback } = require('../../../../utils/logger/fallback/fallback');
const { fallbackHelper } = require('../../../../utils/helpers/fallback/fallback');

describe('Helpers Fallback – loggerFallback (Ultra avancé)', () => {
  it('ajoute un log fallback pour auditabilité', () => {
    const log = loggerFallback('Fallback test');
    expect(log).toHaveProperty('timestamp');
    expect(log.message).toBe('Fallback test');
  });

  it('gère les logs vides', () => {
    const log = loggerFallback('');
    expect(log).toHaveProperty('timestamp');
    expect(log.message).toBe('');
  });
});

describe('Helpers Fallback – fallbackHelper (Ultra avancé)', () => {
  it('retourne le résultat de la fonction si pas d’erreur', () => {
    const fn = () => 42;
    expect(fallbackHelper(fn, 0)).toBe(42);
  });

  it('retourne la valeur par défaut si la fonction échoue', () => {
    const fn = () => { throw new Error('fail'); };
    expect(fallbackHelper(fn, 'defaut')).toBe('defaut');
  });

  it('gère les valeurs nulles et edge cases', () => {
    expect(fallbackHelper(() => null, 'x')).toBe(null);
  });

  it('auditabilité: la sortie est traçable', () => {
    expect(typeof fallbackHelper(() => 'ok', 'ko')).toBe('string');
  });
});
